### PR TITLE
Fix return value usage on error

### PR DESCRIPTION
**Pull request by tschoonj**
WORK IN PROGRESS

@vsoch this is my fix for get. I am not sure what to do if there are multiple images passed: should I abort after the first failure instead?

### CODE REVIEW

**Comment by vsoch** [comment]
I would actually suggest that we change the functions to just be run for one image at a time. There is no real benefit to trying to do multiple at once - the user can just run the function multiple times, and each would correspond with a clean return code.

There are a few spots in the docs that mention being able to do multiple (when interacting with the python client) so we should update those too.

**Comment by tschoonj** [comment]
Ok makes sense. will do that tomorrow

File: sregistry/client/pull.py [hunk 1]
```diff
@@ -18,32 +18,27 @@ def main(args,parser,subparser):
 
     from sregistry.main import get_client
  
-    images = args.image
-    if not isinstance(images, list):
-        images = [images]
+    image = args.image
+    if not isinstance(image, str):
```
Comment (vsoch) [inline sregistry/client/pull.py]: I would instead check if it exists, argparse is going to require it to be defined, and a string.

File: sregistry/client/pull.py [hunk 2]
```diff
@@ -18,32 +18,27 @@ def main(args,parser,subparser):
...
-        if not hasattr(cli, 'pull'):
-            msg = "pull is not implemented for %s. Why don't you add it?"
-            bot.exit(msg % cli.client_name)
-
-        response = cli.pull(images=image,
-                            file_name=name,
-                            force=args.force,
-                            save=do_save)
+    response = cli.pull(images=image,
+                        file_name=name,
```
Comment (vsoch) [inline sregistry/client/pull.py]: The function here - images should be changed to image (unless you think it should remain to support multiple?)

File: sregistry/main/s3/pull.py [hunk 3]
```diff
@@ -60,10 +60,10 @@ def pull(self, images, file_name=None, save=True, **kwargs):
             if e.response['Error']['Code'] == "404":
 
                 # Case 1, image not found, but not error with API
-                bot.error('Cannot find %s!' % name)      
+                bot.error('Cannot find %s!' % file_name)
 
                 # Try to help the user with suggestions
-                results = self._search_all(query=image)
+                results = self._search_all()
```
Comment (vsoch) [inline sregistry/main/s3/pull.py]: Why are you removing the query?

File: sregistry/client/pull.py (hunk 1 above)
Comment (tschoonj) [reply sregistry/client/pull.py]: Good morning!

You're right! Will fix that!

File: sregistry/client/pull.py (hunk 2 above)
Comment (tschoonj) [reply sregistry/client/pull.py]: Sure, I can do that.

File: sregistry/main/s3/pull.py (hunk 3 above)
Comment (tschoonj) [reply sregistry/main/s3/pull.py]: because the query argument isn't supported by the s3 implementation of search_all(). Was getting a TypeError...

File: sregistry/client/pull.py (hunk 1 above)
Comment (vsoch) [reply sregistry/client/pull.py]: Actually if you do this and don’t use nargs the user will need to specify a flag - I don’t think we want that. It’s better to just require 1 argument and grab it from index 0 in the resulting list.

File: sregistry/client/pull.py (hunk 2 above)
Comment (vsoch) [reply sregistry/client/pull.py]: I wanted to bring up for discussion - given the number of pull clients, it might be best to leave as is (I believe multiple or one image are both supported).

File: sregistry/main/s3/pull.py (hunk 3 above)
Comment (vsoch) [reply sregistry/main/s3/pull.py]: Most clients have a search that can support an arg (and if None or “” just search all). The solution is to fix to support that and not remove the argument entirely.

File: sregistry/client/pull.py (hunk 1 above)
Comment (tschoonj) [reply sregistry/client/pull.py]: I don't think a flag is necessary: it just expects a single image name. This is what I get now:

```bash
sregistry pull -h
usage: sregistry pull [-h] [--name NAME] [--force] [--no-cache] image

positional arguments:
  image        full uri of image

optional arguments:
  -h, --help   show this help message and exit
  --name NAME  custom name for image
  --force, -f  force overwrite of existing image
  --no-cache   if storage active, don't add the image to it
```

File: sregistry/client/pull.py (hunk 2 above)
Comment (tschoonj) [reply sregistry/client/pull.py]: Ok

File: sregistry/main/s3/pull.py (hunk 3 above)
Comment (tschoonj) [reply sregistry/main/s3/pull.py]: Certainly, thought I had stumbled upon a bug :smile:

File: sregistry/client/pull.py (hunk 1 above)
Comment (vsoch) [reply sregistry/client/pull.py]: Yeah I think you are right, it’s the — that distinguishes!

It’s too early I’m going to go back to sleep for a bit, looking forward to reviewing further when it’s light out! 🌞

File: sregistry/main/s3/pull.py (hunk 3 above)
Comment (tschoonj) [reply sregistry/main/s3/pull.py]: Had another look at this and I really don't think it makes sense to add the `query` flag here. 
If this line in the code has been reached, the image doesn't exist in the bucket, so what should be done is just list all images that are present in the bucket, which is what `search_all` does.

The only `search_all` option that makes sense to me is `quiet`, which I think should be set to `True`, as currently the bucket images are printed twice.

Also, none of the other `search_all` methods have this kwarg.

Comment (vsoch) [reply sregistry/main/s3/pull.py]: Here is an example of what I'm talking about: https://github.com/singularityhub/sregistry-cli/blob/master/sregistry/main/registry/query.py#L20 The general search function takes a query argument (typically the name of the image) and based on the formatting, will return a call to a sub-search function. In the case of sregistry (the link I showed above) we parse the query to determine if the user is searching for a collection or container, and if it's not, just return search_all() to list all. Since pull is given an image uri, we might consider just handing off the collection name (or some piece of the string) to search here https://github.com/singularityhub/sregistry-cli/blob/master/sregistry/main/s3/query.py#L18 and then returning a subset of results that might match. If you think it's more likely that buckets are small and listing all of the results is really the best option, we could consider that too.

File: sregistry/client/__init__.py [hunk 4]
```diff
@@ -71,24 +71,24 @@ def get_parser():
...
 
     # Get path to an image
     get = subparsers.add_parser("get",
                                 help="get an image path from your storage")
 
-    get.add_argument("query", nargs='*', 
+    get.add_argument("query",
                      help="image search query to inspect", 
-                     type=str, default="*")
+                     type=str, default=None)
```
Comment (vsoch) [inline sregistry/client/__init__.py]: If you add a default as None, then that would need to be checked for. If you don't add a default, would it error without it (as it's a required positional argument?) For some of these I see you added None, and others not, could you explain this decision?

File: sregistry/client/__init__.py [hunk 5]
```diff
@@ -293,10 +293,9 @@ def help(return_code=0):
     # If the user didn't provide any arguments, show the full help
     if len(sys.argv) == 1:
         help()
-    try:
-        args = parser.parse_args()
-    except:
-        sys.exit(0)
+
```
Comment (vsoch) [inline sregistry/client/__init__.py]: Why did you remove this here?

File: sregistry/client/share.py [hunk 6]
```diff
@@ -15,21 +15,17 @@ def main(args, parser, subparser):
...
     from sregistry.main import get_client
-    images = args.image
+    image = args.image
 
-    if not isinstance(images,list):
-        images = [images]
-
-    for image in images:
-        print(image)
+    print(image)
```
Comment (vsoch) [inline sregistry/client/share.py]: This is probably a debug statement that should have been removed (oups).

File: sregistry/main/google_build/delete.py [hunk 7]
```diff
@@ -50,16 +51,22 @@ def delete(self, name):
 
     bot.debug("DELETE %s" % name)
 
+    # this is broken!!!
     for file_object in files:
```
Comment (vsoch) [inline sregistry/main/google_build/delete.py]: Is there a change in the storage API? Could you elaborate more on "this is broken?"

File: sregistry/main/google_storage/delete.py [hunk 8]
```diff
@@ -50,18 +51,22 @@ def delete(self, name):
 
     bot.debug("DELETE %s" % name)
 
+    # broken??
     for file_object in files:
```
Comment (vsoch) [inline sregistry/main/google_storage/delete.py]: Same as above - versions here are important, so here is what I have:

```bash
gapic-google-cloud-datastore-v1==0.15.3
gapic-google-cloud-error-reporting-v1beta1==0.15.3
gapic-google-cloud-logging-v2==0.91.3
google==2.0.1
google-api-core==0.1.4
google-api-python-client==1.7.4
google-auth==1.5.0
google-auth-httplib2==0.0.3
google-cloud==0.33.1
google-cloud-bigquery==0.28.0
google-cloud-bigquery-datatransfer==0.1.1
google-cloud-bigtable==0.28.1
google-cloud-container==0.1.1
google-cloud-core==0.28.1
google-cloud-datastore==1.4.0
google-cloud-dns==0.28.0
google-cloud-error-reporting==0.28.0
google-cloud-firestore==0.28.0
google-cloud-language==1.0.2
google-cloud-logging==1.4.0
google-cloud-monitoring==0.28.1
google-cloud-pubsub==0.30.1
google-cloud-resource-manager==0.28.1
google-cloud-runtimeconfig==0.28.1
google-cloud-spanner==0.29.0
google-cloud-speech==0.30.0
google-cloud-storage==1.6.0
google-cloud-trace==0.17.0
google-cloud-translate==1.3.1
google-cloud-videointelligence==1.0.1
google-cloud-vision==0.29.0
google-gax==0.15.16
google-resumable-media==0.3.1
googleapis-common-protos==1.5.3
grpc-google-iam-v1==0.11.4
proto-google-cloud-datastore-v1==0.90.4
proto-google-cloud-error-reporting-v1beta1==0.15.3
proto-google-cloud-logging-v2==0.91.3
```

File: sregistry/main/s3/delete.py [hunk 9]
```diff
@@ -23,8 +23,24 @@ def remove(self, image, force=False):
     try:
         _object = self.bucket.Object(uri)
         _object.load() # this throws an exception if the object does not exist! -> if delete() fails no exception is thrown...
-        _object.delete()
+    
+        # taken from main.registry.delete.remove
+        continue_delete = True
+        if force is False:
+            response = input("Are you sure you want to delete %s?" % uri)
```
Comment (vsoch) [inline sregistry/main/s3/delete.py]: If you are going to copy paste code, it might be better done as a function that both use.

File: sregistry/main/s3/pull.py (hunk 3 above)
Comment (tschoonj) [reply sregistry/main/s3/pull.py]: Oh I see, so basically all I need to do is use the `main.s3.search` method instead of `main.s3._search_all`?

File: sregistry/client/__init__.py (hunk 4 above)
Comment (tschoonj) [reply sregistry/client/__init__.py]: Actually if you give it no arguments, you already get an appropriate error message:

```
$ sregistry get
usage: sregistry get [-h] query
sregistry get: error: the following arguments are required: query
```

File: sregistry/main/s3/pull.py (hunk 3 above)
Comment (vsoch) [reply sregistry/main/s3/pull.py]: Yes, but if the user got the container name wrong, it might make sense to only search with the collection name. You could try a search with the collection name first, and if there are no results, instruct the user to use the `sregistry search` to list the available containers.

File: sregistry/client/__init__.py (hunk 4 above)
Comment (vsoch) [reply sregistry/client/__init__.py]: This is with the default as None - then is the default needed at all?

Comment (tschoonj) [reply sregistry/client/__init__.py]: It's probably pointless indeed. Will remove it.

File: sregistry/client/__init__.py (hunk 5 above)
Comment (tschoonj) [reply sregistry/client/__init__.py]: Because if you don't catch the exception, you get a clean exit with value 2, which I think is better than returning with value 0.

File: sregistry/client/share.py (hunk 6 above)
Comment (tschoonj) [reply sregistry/client/share.py]: Fixed

File: sregistry/main/google_build/delete.py (hunk 7 above)
Comment (tschoonj) [reply sregistry/main/google_build/delete.py]: Well I didn't test it, but there is no reference to variable `files` before it is used, which looks very suspicious to me...

File: sregistry/main/google_storage/delete.py (hunk 8 above)
Comment (tschoonj) [reply sregistry/main/google_storage/delete.py]: Same problem here

File: sregistry/client/__init__.py (hunk 5 above)
Comment (vsoch) [reply sregistry/client/__init__.py]: okay, great.

File: sregistry/main/google_build/delete.py (hunk 7 above)
Comment (vsoch) [reply sregistry/main/google_build/delete.py]: rofl, I don't even know how I messed that up. I'll take a look.

File: sregistry/main/s3/pull.py (hunk 3 above)
Comment (tschoonj) [reply sregistry/main/s3/pull.py]: Hmmm, would you mind fixing this yourself at some point? This would take more work than I am willing to put in on a friday afternoon :smile:

Comment (vsoch) [reply sregistry/main/s3/pull.py]: Yes of course :)

File: sregistry/main/s3/delete.py (hunk 9 above)
Comment (tschoonj) [reply sregistry/main/s3/delete.py]: Fixed in my last commit

**Comment by tschoonj** [comment]
I think I have put in all the changes I wanted to, let me know if you need me to do anything else!

File: sregistry/main/google_storage/delete.py (hunk 8 above)
Comment (vsoch) [reply sregistry/main/google_storage/delete.py]: I'll handle this in another PR.

**Comment by vsoch** [comment]
The one final detail is to add notes for what you changed in the CHANGELOG.md, and to bump the version.

**Comment by tschoonj** [comment]
I always forget that!

**Comment by vsoch** [comment]
And this is why I'm here, this is my purpose! :)

https://www.youtube.com/watch?v=X7HmltUWXgs

**Comment by vsoch** [comment]
https://pypi.org/project/sregistry/0.2.13/

**Comment by tschoonj** [comment]
Thanks!

By the way, I dont know if you have anything to do with https://www.sylabs.io/singularity/whos-using-singularity/, but I sent an email some time ago asking them to add my [employer's](https://www.diamond.ac.uk/Home.html) logo to the list, but so far nothing has happened... Could you perhaps remind the people managing that website? Not hugely important but we are proud Singularity users here 😄

**Comment by vsoch** [comment]
I was an original developer (and continue to develop) but I'm not employed by Sylabs. I think your best bet would be to ask on slack, or open an issue in the user-docs (that are connected to that site). https://github.com/sylabs/singularity-userdocs

**Comment by tschoonj** [comment]
Ok, thanks will do that!
